lib/class-wp-theme-json.php

```diff
@@ -0,0 +1,953 @@
+	 * @param array  $schema Schema to use for normalization.
+	 */
+	private function process_key( $key, &$input, $schema ) {
+		if ( ! is_array( $input ) || ! array_key_exists( $key, $input ) ) {
+			return;
+		}
+
+		// Consider valid the input value.
+		if ( null === $schema[ $key ] ) {
+			return;
+		}
+
+		if ( ! is_array( $input[ $key ] ) ) {
+			unset( $input[ $key ] );
+			return;
+		}
+
+		$input[ $key ] = array_intersect_key(
+			$input[ $key ],
+			$schema[ $key ]
+		);
+
+		if ( 0 === count( $input[ $key ] ) ) {
+			unset( $input[ $key ] );
+		}
+	}
+
+	/**
+	 * Given a context, it returns its settings subtree.
+	 *
+	 * @param array $context Context adhering to the theme.json schema.
+	 *
+	 * @return array|null The settings subtree.
+	 */
+	private function extract_settings( $context ) {
+		if (
+			! array_key_exists( 'settings', $context ) ||
+			empty( $context['settings'] )
+		) {
+			return null;
+		}
+
+		return $context['settings'];
+	}
+
+	/**
+	 * Given a tree, it creates a flattened one
+	 * by merging the keys and binding the leaf values
+	 * to the new keys.
+	 *
+	 * It also transforms camelCase names into kebab-case
+	 * and substitutes '/' by '-'.
+	 *
+	 * This is thought to be useful to generate
+	 * CSS Custom Properties from a tree,
+	 * although there's nothing in the implementation
+	 * of this function that requires that format.
+	 *
+	 * For example, assuming the given prefix is '--wp'
+	 * and the token is '--', for this input tree:
+	 *
+	 * {
+	 *   'some/property': 'value',
+	 *   'nestedProperty': {
+	 *     'sub-property': 'value'
+	 *   }
+	 * }
+	 *
+	 * it'll return this output:
+	 *
+	 * {
+	 *   '--wp--some-property': 'value',
+	 *   '--wp--nested-property--sub-property': 'value'
+	 * }
+	 *
+	 * @param array  $tree Input tree to process.
+	 * @param string $prefix Prefix to prepend to each variable. '' by default.
+	 * @param string $token Token to use between levels. '--' by default.
+	 *
+	 * @return array The flattened tree.
+	 */
+	private function flatten_tree( $tree, $prefix = '', $token = '--' ) {
+		$result = array();
+		foreach ( $tree as $property => $value ) {
+			$new_key = $prefix . str_replace(
+				'/',
+				'-',
+				strtolower( preg_replace( '/(?<!^)[A-Z]/', '-$0', $property ) ) // CamelCase to kebab-case.
+			);
+
+			if ( is_array( $value ) ) {
+				$new_prefix = $new_key . $token;
+				$result     = array_merge(
+					$result,
+					$this->flatten_tree( $value, $new_prefix, $token )
+				);
+			} else {
+				$result[ $new_key ] = $value;
+			}
+		}
+		return $result;
+	}
+
+	/**
+	 * Utility to extract a path from an array.
+	 *
+	 * @param array $array    Array we want to retrieve data from.
+	 * @param array $path     Array containing the path to retrieve.
+	 * @param array $default  Value to return if $array doesn't contain the $path.
+	 *
+	 * @return array Data at the given $path or $default.
+	 */
+	private function get_from_path( $array, $path, $default = array() ) {
```

Comment:
This function is the dash get equivalent? I guess we can use the existing one and avoid duplication? The function is also used by the server-side supports mechanism and will be part of the core even if internal for now.

Reply:
It is! I wasn't sure what was its status given that you mentioned that it was marked as internal for the block supports, so I wasn't sure it was available any longer. Happy to do what you think is best.

Reply:
It is marked as internal in the core. But I guess we can still use it in Gutenberg. And even in core, it will be public as I think it is a useful function (we just need more code showing its usefulness). I think for now in this PR we should just use the function we already have on Gutenberg and avoid duplicating it.

Reply:
https://github.com/WordPress/gutenberg/pull/26803/commits/aee8a216fd4fdf017d50089b1b96a9c4b32aca78